browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,153 @@
+#include "components/policy/policy_constants.h"
+#include "components/prefs/pref_service.h"
+#include "net/dns/dns_config.h"
+#include "net/dns/public/dns_over_https_config.h"
+#include "net/dns/public/doh_provider_entry.h"
+#include "net/dns/public/secure_dns_mode.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace secure_dns = chrome_browser_net::secure_dns;
+
+namespace brave_vpn {
+
+namespace {
+const char kCloudflareDnsProviderURL[] =
+    "https://chrome.cloudflare-dns.com/dns-query";
+
+void SkipDNSDialog(PrefService* prefs, bool checked) {
+  if (!prefs)
+    return;
+  prefs->SetBoolean(prefs::kBraveVPNShowDNSPolicyWarningDialog, !checked);
+}
+
+std::string GetDoHServers(const std::string* user_servers) {
+  return user_servers && !user_servers->empty() ? *user_servers
+                                                : kCloudflareDnsProviderURL;
+}
+
+gfx::NativeWindow GetAnchorBrowserWindow() {
+  auto* browser = chrome::FindLastActive();
+  return browser ? browser->window()->GetNativeWindow()
+                 : gfx::kNullNativeWindow;
+}
+
+}  // namespace
+
+BraveVpnDnsObserverService::BraveVpnDnsObserverService(
+    PrefService* local_state,
+    PrefService* profile_prefs,
+    DnsPolicyReaderCallback callback)
+    : policy_reader_(std::move(callback)),
+      local_state_(local_state),
+      profile_prefs_(profile_prefs) {
+  DCHECK(profile_prefs_);
+  DCHECK(local_state_);
+}
+
+BraveVpnDnsObserverService::~BraveVpnDnsObserverService() = default;
+
+bool BraveVpnDnsObserverService::IsDnsModeConfiguredByPolicy() const {
+  auto policy_value = (policy_reader_)
+                          ? policy_reader_.Run(policy::key::kDnsOverHttpsMode)
+                          : absl::nullopt;
+  bool doh_policy_set =
+      policy_value.has_value() && !policy_value.value().empty();
+  return doh_policy_set ||
+         SystemNetworkContextManager::GetStubResolverConfigReader()
+             ->ShouldDisableDohForManaged();
+}
+
+void BraveVpnDnsObserverService::ShowPolicyWarningMessage() {
+  if (!profile_prefs_->GetBoolean(prefs::kBraveVPNShowDNSPolicyWarningDialog)) {
+    return;
+  }
+
+  if (policy_callback_) {
+    std::move(policy_callback_).Run();
+    return;
+  }
+
+  chrome::ShowWarningMessageBoxWithCheckbox(
+      GetAnchorBrowserWindow(), l10n_util::GetStringUTF16(IDS_PRODUCT_NAME),
+      l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_POLICY_ALERT),
+      l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_POLICY_CHECKBOX),
+      base::BindOnce(&SkipDNSDialog, profile_prefs_));
+}
+
+void BraveVpnDnsObserverService::ShowVpnNotificationDialog() {
+  if (skip_notification_dialog_for_testing_)
+    return;
+  VpnNotificationDialogView::Show(chrome::FindLastActive());
+}
+
+void BraveVpnDnsObserverService::UnlockDNS() {
+  local_state_->ClearPref(brave_vpn::prefs::kBraveVpnDnsConfig);
+  // Read DNS config to initiate update of actual state.
+  SystemNetworkContextManager::GetStubResolverConfigReader()
```

Comment:
this is fragile, it should be triggered by a pref observer as discussed in dm

Reply:
decided to leave it after discussion in dm